inspirehep/modules/workflows/tasks/manual_merging.py

```diff
@@ -110,16 +110,28 @@ def save_roots(obj, eng):
     """
     def _merge_roots(new_uuid, head_roots, update_roots):
         """Return the new roots to link to head."""
-        head_sources = [h.source for h in head_roots]
-        for u in update_roots:
-            if u.source not in head_sources:
+        head_sources = {h.source: h for h in head_roots}
+        for update_root in update_roots:
+            if update_root.source not in head_sources:
                 head_roots.append(
                     WorkflowsRecordSources(
-                        source=u.source,
+                        source=update_root.source,
                         record_id=new_uuid,
-                        json=u.json
+                        json=update_root.json
                     )
                 )
+            elif head_sources[update_root.source].updated < update_root.updated:
+                head_roots.remove(head_sources[update_root.source])
+                db.session.delete(head_sources[update_root.source])
+                head_roots.append(
+                    WorkflowsRecordSources(
+                        source=update_root.source,
+                        record_id=new_uuid,
+                        json=update_root.json
+                    )
+                )
+            db.session.delete(update_root)
+            db.session.commit()
```

Comment:
do we need the commit here?

Reply:
Yes! Otherwise, when doing the bulk save we get a duplicate key value pk violation. Apparently `db.session.delete` is not enough for the staging area to track the deletions and insertions. You have to first delete the existing record that you want to re-add with the same pk. However, I believe I will try to do a `db.session.merge()` instead for each root.

Reply:
I think the issue is about isolation: is it OK in principle for another process to see a partial merge? if not, it should be done in one transaction.